tests/model/model_flattening.cpp

```diff
@@ -0,0 +1,623 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "gtest/gtest.h"
+
+#include <libcellml>
+
+#include "test_utils.h"
+
+TEST(ModelFlattening, modelWithoutImports)
+{
+    const std::string e =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"units\">\n"
+        "  <units name=\"ms\">\n"
+        "    <unit prefix=\"milli\" units=\"second\"/>\n"
+        "  </units>\n"
+        "</model>\n";
+
+    auto parser = libcellml::Parser::create();
+    auto modelUnitsDefinitions = parser->parseModel(e);
+
+    EXPECT_FALSE(modelUnitsDefinitions->hasUnresolvedImports());
+
+    modelUnitsDefinitions->flatten();
+
+    auto printer = libcellml::Printer::create();
+
+    auto a = printer->printModel(modelUnitsDefinitions);
+    EXPECT_EQ(e, a);
+}
+
+TEST(ModelFlattening, modelWithUnresolvedImports)
+{
+    const std::string modelImportingUnits =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"importer_of_units\">\n"
+        "  <import xlink:href=\"unitsdefinitions.xml\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n"
+        "    <units units_ref=\"ms\" name=\"my_ms\"/>\n"
+        "  </import>\n"
+        "</model>\n";
+
+    auto parser = libcellml::Parser::create();
+    auto modelWithUnitsImports = parser->parseModel(modelImportingUnits);
+
+    EXPECT_TRUE(modelWithUnitsImports->hasUnresolvedImports());
+
+    modelWithUnitsImports->flatten();
+
+    auto printer = libcellml::Printer::create();
+
+    auto a = printer->printModel(modelWithUnitsImports);
+    EXPECT_EQ(modelImportingUnits, a);
+}
+
+TEST(ModelFlattening, importedUnits)
+{
+    const std::string unitsDefiniitonModel =
```

Comment:
these typos should probably be fixed, but will be easier to do them all at once in a proper editor rather than online via Github.com :)

Reply:
Done.